includes/plugins.php

```diff
@@ -157,3 +157,172 @@ function plugin_deactivation_warning() {
+	if ( isset($plugins->response) && is_array($plugins->response) ) {
+		$plugins = array_keys( $plugins->response );
+		foreach ( $plugins as $plugin_file ) {
+			add_action( "after_plugin_row_$plugin_file", __NAMESPACE__ . '\set_custom_update_notification', 10, 2 );
+		}
+	}
+}
+
+/**
+ * Set the custom update notification for plugins which require
+ * updates.
+ *
+ * @param string $file        Plugin basename.
+ * @param array  $plugin_data Plugin information.
+ */
+function set_custom_update_notification( $file, $plugin_data ) {
+	$current = get_site_transient( 'update_plugins' );
+	if ( ! isset( $current->response[ $file ] ) ) {
+		return false;
+	}
+
+	$response = $current->response[ $file ];
+	$plugins_allowedtags = array(
+		'a'       => array( 'href' => array(), 'title' => array() ),
+		'abbr'    => array( 'title' => array() ),
+		'acronym' => array( 'title' => array() ),
+		'code'    => array(),
+		'em'      => array(),
+		'strong'  => array(),
+	);
+
+	/** @var WP_Plugins_List_Table $wp_list_table */
+	$wp_list_table = _get_list_table( 'WP_Plugins_List_Table' );
+
+	if ( is_network_admin() ) {
+		$active_class = is_plugin_active_for_network( $file ) ? ' active' : '';
+	} else {
+		$active_class = is_plugin_active( $file ) ? ' active' : '';
+	}
+
+	printf(
+		'<tr class="plugin-update-tr%s" id="%s" data-slug="%s" data-plugin="%s"><td colspan="%s" class="plugin-update colspanchange"><div class="update-message notice inline notice-warning notice-alt"><p>',
+		esc_attr( $active_class ),
+		esc_attr( $response->slug . '-update' ),
+		esc_attr( $response->slug ),
+		esc_attr( $file ),
+		esc_attr( $wp_list_table->get_column_count() )
+	);
+
+	printf(
+		__( 'Theres is a new version of %s available. <a href="%s" target="_blank">View version %s details</a>.' ),
+		wp_kses( $plugin_data['Name'], $plugins_allowedtags ),
+		esc_url( $plugin_data['PluginURI'] ),
+		esc_html( $response->new_version )
+	);
+
+	print( '</p></div></td></tr>' );
+}
+
+/**
+ * Set the update count in the WP Admin Plugin menu item
+ * when the DISALLOW_FILE_MODS constant is set. This will indicate
+ * when plugins are needing to be updated.
+ *
+ * @global $menu The list of WP Admin menu items.
+ */
+function set_plugin_menu_update_count() {
+	global $menu;
+
+	$menu_index = 65; // wp-admin single site or site on network
+	if ( is_multisite() ) {
+		$menu_index = 20; // wp-admin network settings
+	}
+
+	$update_data = wp_get_update_data();
+	$update_plugins = get_site_transient( 'update_plugins' );
+	if ( ! empty( $update_plugins->response ) ) {
+		$update_data['counts']['plugins'] = count( $update_plugins->response );
+	}
+
+	if ( 1 > $update_data['counts']['plugins'] ) {
+		return;
+	}
+	$count = sprintf(
+		'<span class="update-plugins count-%d"><span class="plugin-count">%d</span></span>',
+		esc_attr( $update_data['counts']['plugins'] ),
+		number_format_i18n( $update_data['counts']['plugins'] )
+	);
+
+	$menu[ $menu_index ][0] = sprintf( __('Plugins %s'), $count );
```

Comment:
I think we should probably be checking that the item in `$menu[ $menu_index ]` is actually the item we expect before replacing items directly in the array this way. Otherwise we could end up with a situation where an item is mis-labeled and non-privileged users would be seeing info they're not supposed to (even if it's just a count).

Reply:
One more gotcha - this match won't work for non-English :) Not a super common issue for our clients, but we should probably do this the right way. I think the best thing to do will be a match against `__( 'Plugins' ) . ' '`, technically it's possible that it wouldn't match because the plugins menu item with the update notification is a separate string but I think that's as sane as we're going to get.

Reply:
Updated - thanks again @helen!